### PR TITLE
Do not expose implicit IInstanceReferenceExpression in an IDelegateCreationExpression targeting static method.

### LINKED ISSUE #15513: IOperation reports 'Instance' for static method access in delegate creation xpression.
From https://github.com/dotnet/roslyn/issues/15325

Given the code:

```c#
using System;

class A
{
    static int Function(int x) => x + x;
    static void Error()
    {
        var func = new Func<int,int>(Function); // <- IDE0009	Member access should be qualified. Lightbulb says to change this to "this.Function", but its compile error.
        func(1);
    }
}
```

For the reference to "Function" (inside ```new Func<int, int>(Function)```), we get an IMemberReferenceExpression that has a non-null ```.Instance``` value.  The kind of this is "InstanceReferenceExpression".  This doesn't make sense as "Function" is a reference to a static method.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #15513.
VSO bug : https://devdiv.visualstudio.com/DevDiv/_workitems/edit/510294

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 1]
```diff
@@ -2858,7 +2858,7 @@ class MyA : System.Attribute
             string expectedOperationTree = @"
 ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.String) (Syntax: 'Prop = ""test""')
   Left: 
-    IPropertyReferenceExpression: System.String MyA.Prop { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.String) (Syntax: 'Prop')
+    IPropertyReferenceExpression: System.String MyA.Prop { get; set; } (OperationKind.PropertyReferenceExpression, Type: System.String) (Syntax: 'Prop')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: TODO: Open an issue for the missing receiver.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs [hunk 2]
```diff
@@ -107,14 +107,14 @@ public void M(int x, string y)
   Initializers(2):
       ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Int32) (Syntax: 'Amount = x')
         Left: 
-          IPropertyReferenceExpression: System.Int32 <anonymous type: System.Int32 Amount, System.String Message>.Amount { get; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'Amount')
+          IPropertyReferenceExpression: System.Int32 <anonymous type: System.Int32 Amount, System.String Message>.Amount { get; } (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'Amount')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: >IPropertyReferenceExpression [](start = 10, length = 28)

TODO: Open an issue for the missing receiver.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs [hunk 3]
```diff
@@ -1724,7 +1724,7 @@ public static void Main()
       Initializers(1):
           ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: X, IsInvalid) (Syntax: 'x = 0')
             Left: 
-              IFieldReferenceExpression: X.x (OperationKind.FieldReferenceExpression, Type: X, IsInvalid) (Syntax: 'x')
+              IFieldReferenceExpression: X.x (Static) (OperationKind.FieldReferenceExpression, Type: X, IsInvalid) (Syntax: 'x')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs]: >IFieldReferenceExpression [](start = 14, length = 25)

TODO: Open an issue about implicit receiver exposed for the static field.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation, @cston Please review.

**Comment by AlekseyTs** [comment]
CC @jinujoseph For ask mode approval.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 4]
```diff
@@ -808,7 +808,15 @@ private IDelegateCreationExpression CreateBoundDelegateCreationExpressionOperati
         private IMethodReferenceExpression CreateBoundMethodGroupSingleMethodOperation(BoundMethodGroup boundMethodGroup, IMethodSymbol methodSymbol, bool suppressVirtualCalls)
         {
             bool isVirtual = (methodSymbol.IsAbstract || methodSymbol.IsOverride || methodSymbol.IsVirtual) && !suppressVirtualCalls;
-            Lazy<IOperation> instance = new Lazy<IOperation>(() => Create(boundMethodGroup.InstanceOpt));
+            BoundExpression receiverOpt = boundMethodGroup.InstanceOpt;
+
+            if (methodSymbol.IsStatic && receiverOpt != null && 
+                receiverOpt.WasCompilerGenerated && receiverOpt.Kind == BoundKind.ThisReference)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Why are we generating a `ThisReference` in the bound tree if the method is static?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: The receiver is coming from a MethodGroup, I guess compiler optimistically synthesizes ```ThisReference``` before overload resolution takes place.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: Is there an intentional change on this line?

**Review by jcouv — APPROVED** [review]
LGTM with nit on header lines (did the encoding change?)

**Comment by natidea** [comment]
Approved to merge via [510294](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/510294)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: > Is there an intentional change on this line?

VS is changing it without by involvement. I believe it is related to the fact that the tool, that IOperation team is using to automatically update baselines, is messing with BOMs. I am not using the tool and instead editing files manually in VS.

**Pull request merged**